src/chips/add_carry_v1.rs

```diff
@@ -0,0 +1,162 @@
+
+            // Previous accumulator amount + new value from a_cell
+            // using binary expression (x_n-4 * 2^16) + (x_n-3 * 2^8) + ... + (x_n * 2)
+            vec![
+                s * ((a + (prev_b * Expression::Constant(Fp::from(1 << 16))) + prev_c)
+                    - ((b * Expression::Constant(Fp::from(1 << 16))) + c)),
+            ]
+        });
+
+        AddCarryConfig {
+            advice: [col_a, col_b, col_c],
+            instance,
+            selector: add_carry_selector,
+        }
+    }
+
+    // Initial accumulator values from instance for expreiment
+    pub fn assign_first_row(
+        &self,
+        mut layouter: impl Layouter<Fp>,
+    ) -> Result<(AssignedCell<Fp, Fp>, AssignedCell<Fp, Fp>), Error> {
+        layouter.assign_region(
+            || "first row",
+            |mut region| {
+                let b_cell = region.assign_advice_from_instance(
+                    || "first acc[1]",
+                    self.config.instance,
+                    0,
+                    self.config.advice[1],
+                    0,
+                )?;
+
+                let c_cell = region.assign_advice_from_instance(
+                    || "first acc[2]",
+                    self.config.instance,
+                    1,
+                    self.config.advice[2],
+                    0,
+                )?;
+
+                Ok((b_cell, c_cell))
+            },
+        )
+    }
+
+    pub fn assign_advice_row(
+        &self,
+        mut layouter: impl Layouter<Fp>,
+        a: Value<Fp>,
+        prev_b: AssignedCell<Fp, Fp>,
+        prev_c: AssignedCell<Fp, Fp>,
+    ) -> Result<(AssignedCell<Fp, Fp>, AssignedCell<Fp, Fp>), Error> {
+        layouter.assign_region(
+            || "adivce row for accumulating",
+            |mut region| {
+                // enable hash selector
+                self.config.selector.enable(&mut region, 1)?;
+
+                let _ = prev_b.copy_advice(|| "prev_b", &mut region, self.config.advice[1], 0);
+                let _ = prev_c.copy_advice(|| "prev_c", &mut region, self.config.advice[2], 0);
+
+                // Assign new amount to the cell inside the region
+                region.assign_advice(|| "a", self.config.advice[0], 1, || a)?;
+
+                // combine accumulated value and new
+                let mut sum = Fp::zero();
+                a.as_ref().map(|f| sum = sum.add(f));
+                prev_b
+                    .value()
+                    .map(|b| sum = sum.add(&b.mul(&Fp::from(1 << 16))));
+                prev_c.value().map(|c| sum = sum.add(c));
+
+                // split by 16bits for two accumulator columns
+                // Alternatives
+                // option1. using additional advice column for calculation
+                // option2. using lookup table for precalulated
+                let max_bits = Fp::from(1 << 16);
+                let split_by_16bits = || {
```

Comment:
Consider moving this support function to Utils

Reply:
Updated at 5ef75a1ad10cdb571bcad59d1144a420b428b7fd